addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusDataThingHandler.java

```diff
@@ -152,16 +153,11 @@ public synchronized void handleCommand(ChannelUID channelUID, Command command) {
         }
 
         if (RefreshType.REFRESH == command) {
-            PollTask pollTask = this.pollTask;
-            if (pollTask == null || manager == null) {
-                logger.debug(
-                        "Thing {} '{}' received REFRESH but no poll task and/or modbus manager is available. Aborting processing of command '{}' to channel '{}'. Not properly initialized or child of endpoint? ",
-                        getThing().getUID(), getThing().getLabel(), command, channelUID);
+            ModbusPollerThingHandler poller = pollerHandler;
+            if (poller == null) {
```

Comment:
As far as I understand you modeling, the `data` thing can be a child of a `poller`, `serial` or `tcp` bridge. Here you only check for a `poller` bridge. And if there is none you ignore the refresh command. So do you have to have a `poller` thing in all cases and the modeling is wrong?

Also I saw that you are using the `ExpressionThreadPoolExecutor`. Since I do not see any scheduling tasks related to for example cron expressions or others, does anything speak against using the `ScheduledExecutorService`, i.e. via `ThreadPoolManager.getScheduledPool(YOUR_POOL_NAME)` which is also the default `scheduler` in the `BaseThingHandler`?

If you only have a few scheduled tasks I would even recommend using this `scheduler` directly and not creating a separate pool for the binding.

Reply:
You have understood correctly that it can be child of `poller`, `serial` or `tcp`. From the [docs (section "Supported things")](https://github.com/ssalonen/openhab2-addons/blob/modbus-openhab2-native-binding/addons/binding/org.openhab.binding.modbus/README.md#supported-things):

> [data thing] Is child of poller (read-only or read-write things) or tcp/serial (write-only things).

So there is no `poller` in all cases. However, `REFRESH` command has an effect only with `data` things that are child of`poller`. This makes sense since if the `data` thing is child of `serial` or `tcp`, it is write-only -- by the very definition, there is nothing to update from the slave.

It is true that I do not use cron-style scheduling as supported by `ExpressionThreadPoolExecutor`. Did not realize there is `ThreadPoolManager.getScheduledPool`... However, it looks like the `getScheduledPool` returns instance of interface`ScheduledExecutorService` which does not have all the methods of `ExpressionThreadPoolExecutor`. Notably, the `purge`, `setRemoveOnCancelPolicy`, `getQueue` & `getActiveCount` methods are all missing. Knowing the implementation one could cast the return value of `getScheduledPool` to concrete class (`ThreadPoolExecutor`) but that would be relying on implementation details, I think. 

Based on the above, I suggest keeping the implementation as is.

I'm not sure what you mean with `BaseThingHandler.scheduler` -- did you notice that all the schduling is _outside_ handlers, in the transport bundle?

Reply:
> I guess I could have introduced different kind of data things for different purposes (read-only, write-only, read-write), but in the end I decided it would simpler for the user to have single thing -- what do you think?

For me it is OK to keep only one `data` thing. But I would leave a comment in the code that it only processes the `REFRESH` command in case it is a child of the `poller` bridge and thus running in a `read` mode. Otherwise developers might stumble across the same question.

The `ExpressionThreadPoolManager` leaks out of the internal ESH code into the API. This will be corrected by the pending PR: https://github.com/eclipse/smarthome/pull/4295

So you should not use this one. Instead you can use the `ThreadPoolManager.getScheduledPool(YOUR_POOL_NAME)` as I suggested. If you schedule a job on them a `ScheduledFuture` is returned which you can store and collect to see how many you have and to query them if they are finished or cancel them. Unfortunately this means a bit more work for your but the `ExpressionThreadPoolManager` will be removed.

> I'm not sure what you mean with BaseThingHandler.scheduler -- did you notice that all the schduling is outside handlers, in the transport bundle?

Sorry forget my comment, you are right, I was just looking at the latest commits and didn't really pay attention to the paths or even bundles where the files were coming from.